packages/cli/src/cmds/agentInstaller/pythonAgentInstaller.ts

```diff
@@ -0,0 +1,120 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable max-classes-per-file */
+
+import { join } from 'path';
+import { promises as fsp } from 'fs';
+import AgentInstaller from './agentInstaller';
+import CommandStruct from './commandStruct';
+import { exists } from '../../utils';
+import chalk from 'chalk';
+
+const REGEX_PKG_DEPENDENCY = /^\s*appmap\s*[=>~]+=.*$/m;
+// include .dev0 so pip will allow prereleases
+const PKG_DEPENDENCY = 'appmap>=1.1.0.dev0';
+
+export class PoetryInstaller implements AgentInstaller {
+  constructor(readonly path: string) {}
+
+  get name(): string {
+    return 'poetry';
+  }
+
+  get buildFile(): string {
+    return 'poetry.lock';
+  }
+
+  get buildFilePath(): string {
+    return join(this.path, this.buildFile);
+  }
+
+  postInstallMessage(): string {
+    return [
+      `Run your tests with ${chalk.blue('APPMAP=true')} in the environment.`,
+      `By default, AppMap files will be output to ${chalk.blue('tmp/appmap')}.`,
+    ].join('\n');
+  }
+
+  async available(): Promise<boolean> {
+    return await exists(this.buildFilePath);
+  }
+
+  verifyCommand(): CommandStruct {
+    return new CommandStruct(
+      'poetry',
+      ['add', '--dev', '--allow-prereleases', 'appmap'],
+      this.path
+    );
+  }
+
+  initCommand(): CommandStruct {
+    return new CommandStruct('poetry', ['run', 'appmap-agent-init'], this.path);
+  }
+
+  installAgent(): void {
```

Comment:
@dustinbyrne this isn't right: there's only one step in the installation of the python agent in a poetry project (the `poetry add` command above).

This is the reason I introduced the notion of a workflow based on the installer type. Any thoughts on how you'd like to address this?

Reply:
Where was the logic for the `poetry` install?

Reply:
`poertry add` updates `pyproject.toml`, and actually installs the `appmap-python` package. Nothing else needs to be done.

Reply:
Ah - the verify command isn't actually run anywhere. Is `verify` the same as `validate`? It makes more sense to me that the install step is responsible for running `poetry add`.

Reply:
No, `verify` isn't the same as `validate` (i.e. the step that calls out to `appmap-agent-validate`).

FWIW, @kgilpin originally came up with these steps, based on the requirements for Ruby and Java projects. They didn't map well to poetry projects. Structure them however seems best to you, based on your new framework. As long as `poetry add` gets called in a poetry project to install the agent, things should be fine.